# Regression

In [1]:
import pandas as pd

# Load data for all 5 cities
path = "../data/"
beijing_df = pd.read_csv(path + "Beijing.txt")
chengdu_df = pd.read_csv(path + "Chengdu.txt")
guangzhou_df = pd.read_csv(path + "Guangzhou.txt")
shanghai_df = pd.read_csv(path + "Shanghai.txt")
shenyang_df = pd.read_csv(path + "Shenyang.txt")

In [2]:
beijing_df
chengdu_df
guangzhou_df
shanghai_df
shenyang_df

year  month  day  hour  season     PM  DEWP   HUMI    PRES  TEMP  \
0      2010      1    1     0       4    NaN -26.0  69.79  1024.0 -22.0   
1      2010      1    1     1       4    NaN -26.0  76.26  1024.0 -23.0   
2      2010      1    1     2       4    NaN -27.0  69.56  1023.0 -23.0   
3      2010      1    1     3       4    NaN -27.0  69.56  1023.0 -23.0   
4      2010      1    1     4       4    NaN -27.0  69.56  1022.0 -23.0   
...     ...    ...  ...   ...     ...    ...   ...    ...     ...   ...   
52579  2015     12   31    19       4  166.0 -10.0  92.42  1031.0  -9.0   
52580  2015     12   31    20       4  259.0 -10.0  79.10  1030.0  -7.0   
52581  2015     12   31    21       4  368.0 -10.0  79.10  1030.0  -7.0   
52582  2015     12   31    22       4  319.0 -10.0  79.10  1028.0  -7.0   
52583  2015     12   31    23       4  275.0  -9.0  79.26  1028.0  -6.0   

           Iws  precipitation  Iprec  
0       1.0289            NaN    NaN  
1       2.5722            NaN    NaN  
2       5.1444            NaN    NaN  
3       7.7166            NaN    NaN  
4       9.7744            NaN    NaN  
...        ...            ...    ...  
52579   2.0000            0.0    0.0  
52580   5.0000            0.0    0.0  
52581   8.0000            0.0    0.0  
52582  11.0000            NaN    NaN  
52583  12.0000            0.0    0.0  

[52584 rows x 13 columns]

In [3]:
# Aggregate each city's data frame to get daily averages (rounded one decimal place) without averaged hour feature
beijing_df_by_day = beijing_df.groupby(['year', 'month', 'day']).mean().round(1)
beijing_df_by_day = beijing_df_by_day.drop('hour', axis=1)
beijing_df_by_day[59:90]
chengdu_df_by_day = chengdu_df.groupby(['year', 'month', 'day']).mean().round(1)
chengdu_df_by_day = chengdu_df_by_day.drop('hour', axis=1)
chengdu_df_by_day[59:90]
guangzhou_df_by_day = guangzhou_df.groupby(['year', 'month', 'day']).mean().round(1)
guangzhou_df_by_day = guangzhou_df_by_day.drop('hour', axis=1)
guangzhou_df_by_day[179:210]
shanghai_df_by_day = shanghai_df.groupby(['year', 'month', 'day']).mean().round(1)
shanghai_df_by_day = shanghai_df_by_day.drop('hour', axis=1)
shanghai_df_by_day[270:307]
shenyang_df_by_day = shenyang_df.groupby(['year', 'month', 'day']).mean().round(1)
shenyang_df_by_day = shenyang_df_by_day.drop('hour', axis=1)
shenyang_df_by_day[209:234]

season  PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                              
2010 7     29      2.0 NaN  21.8  82.4   998.6  25.1   4.5            0.9   
           30      2.0 NaN  23.0  88.7   999.4  25.0   7.4            0.1   
           31      2.0 NaN  22.3  91.7  1000.6  23.7   7.9            0.4   
     8     1       2.0 NaN  15.7  70.7  1007.8  22.2   4.7            0.0   
           2       2.0 NaN  16.2  69.3  1008.7  22.7  10.6            0.0   
           3       2.0 NaN  19.5  77.8  1002.9  23.9   8.1            0.0   
           4       2.0 NaN  21.0  78.9  1003.6  25.0  10.2            0.0   
           5       2.0 NaN  20.9  89.2  1005.5  22.8   6.8            2.9   
           6       2.0 NaN  17.6  77.8  1014.3  22.2   8.1            0.0   
           7       2.0 NaN  17.3  74.9  1012.5  22.6   4.8            0.0   
           8       2.0 NaN  19.8  90.8  1005.5  21.4   6.1            1.4   
           9       2.0 NaN  21.1  88.8  1005.5  23.1   4.9            0.1   
           10      2.0 NaN  21.6  82.8  1006.6  24.8   3.2            0.0   
           11      2.0 NaN  21.0  82.2  1004.5  24.5   4.5            0.0   
           12      2.0 NaN  19.5  74.4  1001.6  25.0   9.1            0.0   
           13      2.0 NaN  21.6  84.4  1001.1  24.5  15.0            0.0   
           14      2.0 NaN  21.8  91.8   999.8  23.2   3.1            0.5   
           15      2.0 NaN  19.2  79.2  1002.6  23.3  14.2            0.0   
           16      2.0 NaN  17.8  71.2  1007.8  23.9   7.0            0.0   
           17      2.0 NaN  19.0  73.5  1009.7  24.5  10.6            0.0   
           18      2.0 NaN  18.0  66.8  1013.0  25.2  10.5            0.0   
           19      2.0 NaN  22.4  79.9  1013.1  26.3  21.7            0.0   
           20      2.0 NaN  22.5  89.1  1014.6  24.5  21.8            0.2   
           21      2.0 NaN  20.2  92.2  1012.4  21.5   8.2            4.0   
           22      2.0 NaN  17.2  92.7  1010.7  18.4   5.1            1.8   

                Iprec  
year month day         
2010 7     29     4.6  
           30     0.2  
           31     1.3  
     8     1      0.0  
           2      0.0  
           3      0.0  
           4      0.0  
           5     24.6  
           6      0.0  
           7      0.0  
           8      7.9  
           9      0.2  
           10     0.0  
           11     0.0  
           12     0.0  
           13     0.0  
           14     1.3  
           15     0.0  
           16     0.0  
           17     0.0  
           18     0.0  
           19     0.0  
           20     0.4  
           21    15.4  
           22    67.6

## 1.1 Answer the following questions

- Which of the 5 cities has the largest temperature range (e.g. highest temperature - lowest temperature) during the period the dataset was collected?

In [4]:
# Get max temp - min temp for each city.
beijing_range = beijing_df['TEMP'].max() - beijing_df['TEMP'].min()
beijing_range

61.0

In [5]:
chengdu_range = chengdu_df['TEMP'].max() - chengdu_df['TEMP'].min()
chengdu_range

41.0

In [6]:
import numpy as np
guangzhou_range = guangzhou_df['TEMP'].max() - guangzhou_df['TEMP'].min()
np.round(guangzhou_range, 1)

35.7

In [7]:
shanghai_range = shanghai_df['TEMP'].max() - shanghai_df['TEMP'].min()
shanghai_range

46.0

In [8]:
shenyang_range = shenyang_df['TEMP'].max() - shenyang_df['TEMP'].min()
shenyang_range

63.0

**Shenyang is the city with the largest temperature range (63) during the period the dataset was collected because it is the city with the largest difference of max temp minus min temp.**

- Which cities would you consider to be the most and least polluted? Explain the logic for your answer in detail.

In [9]:
# Average PM for each city.
np.round(beijing_df['PM'].mean(), 1)

95.9

In [10]:
np.round(chengdu_df['PM'].mean(), 1)

83.4

In [11]:
np.round(guangzhou_df['PM'].mean(), 1)

50.8

In [12]:
np.round(shanghai_df['PM'].mean(), 1)

52.9

In [13]:
np.round(shenyang_df['PM'].mean(), 1)

75.8

**I would consider Beijing the most polluted city and Guangzhou the least polluted city. I got the average PM2.5 concentration (PM feature) for each city, chosen because this feature describes atmospheric particulate matter and is therefor the feature most closely related to pollution. I supposed that the city with the largest average (Beijing) is the most polluted, and the city with the lowest average (Guangzhou) is the least polluted.**

- What is the average temperature for each of the four seasons in each of the cities?

In [14]:
# Create pivot table indexed by season, get mean temp
beijing_seasons = beijing_df.pivot_table('TEMP', index='season', aggfunc=np.mean).round(1)
beijing_seasons.head()

TEMP
season      
1       14.0
2       25.9
3       12.6
4       -2.4

In [15]:
chengdu_season = chengdu_df.pivot_table('TEMP', index='season', aggfunc=np.mean).round(1)
chengdu_season.head()

TEMP
season      
1       18.3
2       26.0
3       18.4
4        8.1

In [16]:
guangzhou_season = guangzhou_df.pivot_table('TEMP', index='season', aggfunc=np.mean).round(1)
guangzhou_season.head()

TEMP
season      
1.0     21.9
2.0     28.2
3.0     23.6
4.0     14.1

In [17]:
shanghai_season = shanghai_df.pivot_table('TEMP', index='season', aggfunc=np.mean).round(1)
shanghai_season.head()

TEMP
season      
1       16.0
2       27.7
3       19.8
4        6.2

In [18]:
shenyang_season = shenyang_df.pivot_table('TEMP', index='season', aggfunc=np.mean).round(1)
shenyang_season.head()

TEMP
season      
1        9.5
2       23.6
3        9.8
4       -9.2

- What are the hottest summer and coldest winters observed? Explain how you defined hottest and coldest?

**I will define hottest summer as the summer (by year) having the most instances of average high temperature and humidity in the top 10. I chose this because a max temperature could be an outlier and not as representative, and I chose both temperature and humidity because humidity intensifies the feeling of being hot. For coldest winter, I will use only temperature as there is debate regarding how humidity affects the feeling of being cold. The other measure is the same: the winter (by year) having the most instances of average low temperatures in the top 10.** 

In [19]:
# Filter based on seasons == 2.0 (summer)
beijing_summer = beijing_df_by_day['season'] == 2.0
beijing_summer[195:245]
beijing_summers = beijing_df_by_day.loc[beijing_summer]
beijing_summers[131:200]

season     PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                                 
2011 7     10      2.0   87.0  19.2  61.9  1006.5  27.2   6.7            0.0   
           11      2.0  191.8  21.8  74.0  1007.6  27.2  13.7            0.0   
           12      2.0  111.3  21.5  73.3  1008.0  26.8  13.2            0.0   
           13      2.0   93.7  22.0  73.7  1005.3  27.2  18.1            0.0   
           14      2.0  103.2  21.9  75.2  1005.4  26.8   6.2            0.1   
...                ...    ...   ...   ...     ...   ...   ...            ...   
2012 6     12      2.0   20.3   9.2  40.6   996.5  24.3   6.4            0.0   
           13      2.0   70.1  13.5  58.2   999.9  22.3   7.5            0.0   
           14      2.0    9.8  13.1  59.6  1004.0  21.9  10.1            0.0   
           15      2.0   11.1  11.8  51.0  1006.3  23.6   8.7            0.0   
           16      2.0   25.1  12.6  48.3  1004.5  25.8  14.2            0.0   

                Iprec  
year month day         
2011 7     10     0.0  
           11     0.0  
           12     0.0  
           13     0.0  
           14     0.2  
...               ...  
2012 6     12     0.0  
           13     0.0  
           14     0.0  
           15     0.0  
           16     0.0  

[69 rows x 9 columns]

In [20]:
# Get the top 10 hottest summer days
beijing_hottest = beijing_summers.sort_values(by=['TEMP', 'HUMI'], ascending=False)
beijing_hottest[0:10]

season     PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                                 
2010 7     6       2.0   25.0   9.7  26.7  1000.2  32.9  13.8            0.0   
2015 7     13      2.0   43.5  16.3  43.3   997.5  32.4   5.7            0.0   
2010 7     5       2.0   46.3  12.3  37.8   999.6  32.4  15.0            0.0   
2014 7     19      2.0   61.2  23.2  62.1  1007.9  31.7  12.2            0.0   
           20      2.0   60.5  23.2  61.1  1004.8  31.7  29.0            0.0   
2010 7     4       2.0   92.9  19.1  51.9   999.6  31.0   9.3            0.0   
2013 8     17      2.0   38.9  14.8  42.3  1000.6  31.0  45.0            0.0   
2014 7     21      2.0   24.9  19.6  52.4  1001.8  30.8  28.1            0.0   
2010 7     28      2.0  177.0  24.0  68.9   999.5  30.6  49.1            0.0   
           27      2.0  182.7  23.8  68.4  1004.4  30.5  23.5            0.0   

                Iprec  
year month day         
2010 7     6      0.0  
2015 7     13     0.0  
2010 7     5      0.0  
2014 7     19     0.0  
           20     0.0  
2010 7     4      0.0  
2013 8     17     0.0  
2014 7     21     0.0  
2010 7     28     0.0  
           27     0.0

In [21]:
# Filter based on seasons == 4.0 (winter)
beijing_winter = beijing_df_by_day['season'] == 4.0
beijing_winter[325:365]
beijing_winters = beijing_df_by_day.loc[beijing_winter]
beijing_winters[131:200]

season     PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                                 
2011 2     11      4.0   58.0 -12.3  56.7  1030.4  -3.8  28.9            0.0   
           12      4.0   29.4 -17.2  41.5  1032.5  -6.0  11.1            0.0   
           13      4.0   72.0  -8.1  79.7  1030.3  -4.9  10.7            0.1   
           14      4.0   36.9 -17.1  40.6  1032.1  -5.4   9.2            0.0   
           15      4.0  211.9 -10.8  69.0  1027.2  -5.8   2.5            0.0   
...                ...    ...   ...   ...     ...   ...   ...            ...   
2012 1     16      4.0    NaN  -9.8  67.4  1020.8  -4.7   3.5            0.0   
           17      4.0  327.6  -7.0  69.5  1022.5  -2.1   2.3            0.0   
           18      4.0  402.3  -6.4  76.7  1025.6  -2.8   6.2            0.0   
           19      4.0  428.5  -6.7  72.5  1028.8  -2.4   3.6            0.0   
           20      4.0   39.4 -18.9  31.1  1035.0  -3.5  59.3            0.0   

                Iprec  
year month day         
2011 2     11     0.0  
           12     0.0  
           13     0.5  
           14     0.0  
           15     0.0  
...               ...  
2012 1     16     0.0  
           17     0.0  
           18     0.0  
           19     0.0  
           20     0.0  

[69 rows x 9 columns]

In [22]:
beijing_coldest = beijing_winters.sort_values(by='TEMP')
beijing_coldest[0:10]

season     PM  DEWP  HUMI    PRES  TEMP    Iws  precipitation  \
year month day                                                                  
2010 1     5       4.0   43.5 -24.6  42.0  1033.6 -14.5   56.9            0.0   
           6       4.0   59.4 -23.7  39.2  1033.8 -12.5   18.5            0.0   
           7       4.0   72.5 -21.2  49.0  1034.1 -12.5   10.2            0.0   
           12      4.0   20.5 -23.5  39.1  1030.7 -12.4   60.4            0.0   
           8       4.0  174.3 -17.1  64.5  1028.0 -11.7    2.0            0.0   
2012 12    26      4.0    NaN -17.1  64.0  1035.8 -11.5    1.9            0.0   
           24      4.0    NaN -19.6  54.3  1028.2 -11.5   10.5            0.0   
2010 1     4       4.0   29.3 -20.9  46.2  1029.3 -11.5  111.2            0.0   
2012 12    23      4.0   11.8 -23.9  34.4  1036.3 -10.8  128.6            0.0   
2013 1     4       4.0   85.5 -21.1  45.3  1032.8 -10.7   12.8            0.0   

                Iprec  
year month day         
2010 1     5      0.0  
           6      0.0  
           7      0.0  
           12     0.0  
           8      0.0  
2012 12    26     0.0  
           24     0.0  
2010 1     4      0.0  
2012 12    23     0.0  
2013 1     4      0.0

**The hottest summer in Beijing for this dataset was summer of 2010 because it has the most instances (5) in the top 10 hottest summer days. The coldest winter in Beijing for this dataset was the winter of 2010 because it has 6 instances in the top 10 coldest winter days, more than any other winter.
Note: I used the dataframe with daily aggregates because hourly data isn't necessary - and is distracting - for this measure.**

In [23]:
# Filter based on seasons == 2.0 (summer)
chengdu_summer = chengdu_df_by_day['season'] == 2.0
chengdu_summer[195:245]
chengdu_summers = chengdu_df_by_day.loc[chengdu_summer]
chengdu_summers[131:200]

season  PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                              
2011 7     10      2.0 NaN  20.6  83.1  1008.6  23.8   9.6            0.0   
           11      2.0 NaN  18.8  88.8  1010.9  20.7   3.1            0.1   
           12      2.0 NaN  18.8  91.2  1011.4  20.3   5.0            0.3   
           13      2.0 NaN  18.8  84.5  1011.2  21.7   3.3            0.0   
           14      2.0 NaN  20.0  79.0  1010.6  24.1   5.6            0.0   
...                ...  ..   ...   ...     ...   ...   ...            ...   
2012 6     12      2.0 NaN  17.5  60.9  1005.2  26.0  13.2            0.0   
           13      2.0 NaN  18.4  65.0  1003.8  26.2   7.0            0.0   
           14      2.0 NaN  18.2  60.6  1006.2  27.1   2.5            0.0   
           15      2.0 NaN  19.3  68.3  1007.4  25.8   4.6            0.2   
           16      2.0 NaN  18.0  85.1  1010.4  20.6   5.5            0.1   

                Iprec  
year month day         
2011 7     10     0.1  
           11     0.1  
           12     1.7  
           13     0.0  
           14     0.0  
...               ...  
2012 6     12     0.0  
           13     0.1  
           14     0.0  
           15     0.8  
           16     0.6  

[69 rows x 9 columns]

In [24]:
# Get the top 10 hottest summer days
chengdu_hottest = chengdu_summers.sort_values(by=['TEMP', 'HUMI'], ascending=False)
chengdu_hottest[0:10]

season     PM  DEWP  HUMI    PRES  TEMP  Iws  precipitation  \
year month day                                                                
2013 8     24      2.0   61.6  20.3  54.9  1004.9  31.6  5.7            0.0   
2010 7     30      2.0    NaN  25.7  73.3  1000.2  31.4  1.7            1.1   
2015 8     1       2.0   37.8  20.8  57.3  1005.9  31.2  3.7            0.0   
           2       2.0   32.3  20.6  56.4  1005.2  31.2  2.7            0.8   
2010 8     12      2.0    NaN  25.3  73.2  1002.5  30.9  2.9            0.0   
2013 8     23      2.0   86.4  21.1  59.9  1004.3  30.8  3.3            0.0   
2010 7     29      2.0    NaN  25.1  73.9  1000.8  30.7  2.3            0.0   
2012 8     11      2.0  107.3  24.3  71.3  1002.7  30.6  3.5            0.1   
           12      2.0   99.4  23.9  70.0  1001.5  30.4  5.1            0.0   
2011 7     26      2.0    NaN  24.1  70.6  1002.6  30.3  0.8            0.0   

                Iprec  
year month day         
2013 8     24     0.0  
2010 7     30     1.2  
2015 8     1      0.0  
           2      2.0  
2010 8     12     0.0  
2013 8     23     0.0  
2010 7     29     0.0  
2012 8     11     0.1  
           12     0.0  
2011 7     26     0.0

In [25]:
# Filter based on seasons == 4.0 (winter)
chengdu_winter = chengdu_df_by_day['season'] == 4.0
chengdu_winter[325:365]
chengdu_winters = chengdu_df_by_day.loc[chengdu_winter]
chengdu_winters[131:200]

season  PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                              
2011 2     11      4.0 NaN   5.5  83.0  1021.9   8.2  10.1            0.1   
           12      4.0 NaN   4.1  85.6  1020.0   6.4  16.8            0.0   
           13      4.0 NaN   3.7  77.7  1023.5   7.5   2.9            0.0   
           14      4.0 NaN   4.3  84.5  1019.9   6.8   4.5            0.0   
           15      4.0 NaN   4.4  86.8  1017.3   6.5   3.8            0.1   
...                ...  ..   ...   ...     ...   ...   ...            ...   
2012 1     16      4.0 NaN   4.5  83.9  1013.1   7.1   4.3            0.0   
           17      4.0 NaN   2.8  72.7  1014.0   7.8   6.6            0.0   
           18      4.0 NaN   3.4  79.9  1015.4   6.7   3.6            0.0   
           19      4.0 NaN   3.6  81.9  1017.5   6.6   8.2            0.0   
           20      4.0 NaN   3.0  86.7  1018.3   5.0   9.0            0.0   

                Iprec  
year month day         
2011 2     11     0.3  
           12     0.2  
           13     0.0  
           14     0.2  
           15     0.2  
...               ...  
2012 1     16     0.0  
           17     0.0  
           18     0.0  
           19     0.0  
           20     0.0  

[69 rows x 9 columns]

In [26]:
chengdu_coldest = chengdu_winters.sort_values(by='TEMP')
chengdu_coldest[0:10]

season     PM  DEWP  HUMI    PRES  TEMP  Iws  precipitation  \
year month day                                                                
2011 1     19      4.0    NaN  -0.9  84.2  1024.2   1.5  5.5            0.0   
           12      4.0    NaN  -1.6  79.5  1020.0   1.7  4.2            0.0   
2013 1     6       4.0  134.8  -0.2  87.8  1023.7   1.8  2.4            0.0   
2011 1     17      4.0    NaN   0.8  91.5  1025.6   2.0  8.3            0.1   
2010 12    16      4.0    NaN  -1.9  76.2  1031.4   2.2  3.5            0.0   
2011 1     18      4.0    NaN  -0.0  84.5  1022.7   2.3  8.0            0.1   
2013 1     4       4.0  105.5   0.5  86.9  1024.5   2.4  2.6            0.0   
           7       4.0  161.3  -0.1  85.8  1022.1   2.4  2.1            0.0   
2011 1     20      4.0    NaN  -1.9  74.9  1025.2   2.5  4.0            0.0   
           21      4.0    NaN  -1.5  74.3  1023.9   2.8  3.0            0.0   

                Iprec  
year month day         
2011 1     19     0.0  
           12     0.0  
2013 1     6      0.0  
2011 1     17     0.6  
2010 12    16     0.0  
2011 1     18     0.3  
2013 1     4      0.0  
           7      0.0  
2011 1     20     0.0  
           21     0.0

**The hottest summer in Chengdu for this dataset was summer of 2010 because it has the most instances (3) in the top 10 hottest summer days. The coldest winter in Chendgu for this dataset was the winter of 2011 because it has 6 instances in the top 10 coldest winter days, more than any other winter.**

In [27]:
# Filter based on seasons == 2.0 (summer)
guangzhou_summer = guangzhou_df_by_day['season'] == 2.0
guangzhou_summer[195:245]
guangzhou_summers = guangzhou_df_by_day.loc[guangzhou_summer]
guangzhou_summers[131:200]

season    PM  DEWP  HUMI   PRES  TEMP  Iws  precipitation  \
year month day                                                              
2011 7     10      2.0   NaN  24.8  81.3  994.1  28.6  3.6            1.9   
           11      2.0   NaN  24.9  85.8  993.4  27.7  3.2            0.6   
           12      2.0   NaN  25.5  92.1  993.6  26.9  3.2            0.9   
           13      2.0   NaN  25.0  81.8  994.2  28.6  5.9            0.0   
           14      2.0   NaN  24.7  82.7  994.3  28.1  8.2            0.9   
...                ...   ...   ...   ...    ...   ...  ...            ...   
2012 6     12      2.0  43.2  25.2  85.0  991.7  28.1  3.5            0.3   
           13      2.0  52.1  24.7  84.7  992.7  27.6  5.9            0.0   
           14      2.0  67.3  23.5  79.1  992.9  27.7  4.7            0.0   
           15      2.0  79.0  24.3  82.0  992.5  27.9  3.5            1.3   
           16      2.0  33.6  24.0  85.5  992.7  26.8  9.8            0.2   

                Iprec  
year month day         
2011 7     10     1.9  
           11     5.1  
           12     2.3  
           13     0.0  
           14     1.8  
...               ...  
2012 6     12     0.6  
           13     0.0  
           14     0.0  
           15     2.0  
           16     0.2  

[69 rows x 9 columns]

In [28]:
# Get the top 10 hottest summer days
guangzhou_hottest = guangzhou_summers.sort_values(by=['TEMP', 'HUMI'], ascending=False)
guangzhou_hottest[0:10]

season    PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                                
2010 7     5       2.0   NaN  22.9  59.4  1000.9  32.2   6.9            0.0   
           4       2.0   NaN  23.7  62.5  1002.7  32.1   4.0            0.0   
           6       2.0   NaN  24.0  64.8  1001.0  31.7  12.7            0.0   
           8       2.0   NaN  23.7  63.8  1002.5  31.7   4.4            0.0   
           7       2.0   NaN  24.1  66.5  1002.5  31.3   6.4            0.0   
           3       2.0   NaN  23.0  62.3  1003.0  31.3   4.3            0.0   
           2       2.0   NaN  23.0  63.2  1002.5  31.2   4.0            0.0   
2015 7     2       2.0  20.4  24.7  70.1   994.0  31.1  16.3            0.0   
2010 8     10      2.0   NaN  23.7  66.1  1001.0  31.1   6.9            0.0   
           14      2.0   NaN  23.4  64.8  1003.9  31.1   3.1            0.0   

                Iprec  
year month day         
2010 7     5      0.0  
           4      0.0  
           6      0.0  
           8      0.0  
           7      0.0  
           3      0.0  
           2      0.0  
2015 7     2      0.0  
2010 8     10     0.0  
           14     0.0

In [29]:
# Filter based on seasons == 4.0 (winter)
guangzhou_winter = guangzhou_df_by_day['season'] == 4.0
guangzhou_winter[325:365]
guangzhou_winters = guangzhou_df_by_day.loc[guangzhou_winter]
guangzhou_winters[131:200]

season  PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                              
2011 2     11      4.0 NaN   7.4  64.0  1009.7  14.1  10.4            0.0   
           12      4.0 NaN   5.8  69.3  1012.2  11.3  11.1            0.0   
           13      4.0 NaN   7.9  91.3  1010.6   9.3   7.8            1.1   
           14      4.0 NaN   2.7  63.5  1013.7   9.6   9.6            0.0   
           15      4.0 NaN   5.3  88.1  1013.7   7.4  15.8            0.5   
...                ...  ..   ...   ...     ...   ...   ...            ...   
2012 1     16      4.0 NaN  10.4  83.9  1005.4  13.2   9.0            0.1   
           17      4.0 NaN   9.4  77.4  1007.4  13.4  15.7            0.0   
           18      4.0 NaN  13.1  75.5  1005.6  17.6   6.8            0.0   
           19      4.0 NaN  16.0  91.1  1006.3  17.4   5.6            0.1   
           20      4.0 NaN  12.3  90.6  1007.5  13.8   6.1            0.0   

                Iprec  
year month day         
2011 2     11     0.0  
           12     0.0  
           13     6.6  
           14     0.0  
           15     1.8  
...               ...  
2012 1     16     3.8  
           17     0.0  
           18     0.0  
           19     0.2  
           20     0.0  

[69 rows x 9 columns]

In [30]:
guangzhou_coldest = guangzhou_winters.sort_values(by='TEMP')
guangzhou_coldest[0:10]

season    PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                                
2014 2     12      4.0  51.3   1.5  79.5  1011.8   4.9  11.8            0.1   
           11      4.0  41.4  -3.3  55.0  1013.2   5.0   8.2            0.0   
2012 1     26      4.0   NaN   3.7  90.2  1013.3   5.1  15.8            0.0   
2014 2     13      4.0  36.0   2.9  85.2  1014.8   5.3   6.8            0.2   
2012 1     5       4.0   NaN   3.4  87.8  1017.1   5.4   8.1            0.4   
2011 1     12      4.0   NaN   1.0  73.0  1012.3   5.6  10.0            0.2   
2012 1     24      4.0   NaN   4.5  92.0  1012.6   5.7  15.1            0.1   
2011 1     11      4.0   NaN  -0.0  69.0  1012.6   5.8  16.1            0.1   
2012 1     25      4.0   NaN   1.7  74.2  1014.8   6.1  23.8            0.1   
2011 1     4       4.0   NaN   3.8  85.2  1014.1   6.2   6.2            0.7   

                Iprec  
year month day         
2014 2     12     0.4  
           11     0.0  
2012 1     26     0.0  
2014 2     13     0.7  
2012 1     5      2.3  
2011 1     12     2.4  
2012 1     24     0.3  
2011 1     11     0.3  
2012 1     25     0.3  
2011 1     4      4.5

**The hottest summer in Guangzhou for this dataset was summer of 2010 because it has the most instances (9) in the top 10 hottest summer days. The coldest winter in Guangzhou for this dataset was the winter of 2012 because it has 4 instances in the top 10 coldest winter days, more than any other winter.**

In [31]:
# Filter based on seasons == 2.0 (summer)
shanghai_summer = shanghai_df_by_day['season'] == 2.0
shanghai_summer[195:245]
shanghai_summers = shanghai_df_by_day.loc[shanghai_summer]
shanghai_summers[131:200]

season    PM  DEWP  HUMI    PRES  TEMP    Iws  precipitation  \
year month day                                                                 
2011 7     10      2.0   NaN  23.5  73.3  1006.1  28.9   83.2            0.0   
           11      2.0   NaN  23.4  73.6  1006.1  28.7  238.7            0.0   
           12      2.0   NaN  23.5  74.5  1005.3  28.5  381.6            0.0   
           13      2.0   NaN  23.0  70.8  1004.9  29.0  532.7            0.0   
           14      2.0   NaN  23.9  78.3  1004.7  28.2  296.3            0.2   
...                ...   ...   ...   ...     ...   ...    ...            ...   
2012 6     12      2.0  55.8  21.2  77.8   999.9  25.5    7.8            0.0   
           13      2.0  58.2  19.6  71.7  1003.8  25.5   19.1            0.0   
           14      2.0  37.4  20.8  74.5  1005.6  25.8  123.4            0.0   
           15      2.0  21.0  21.3  77.2  1004.5  25.7   47.4            0.0   
           16      2.0  25.9  19.0  71.2  1007.2  24.8   15.4            0.0   

                Iprec  
year month day         
2011 7     10     0.0  
           11     0.0  
           12     0.0  
           13     0.0  
           14     0.7  
...               ...  
2012 6     12     0.0  
           13     0.0  
           14     0.0  
           15     0.0  
           16     0.0  

[69 rows x 9 columns]

In [32]:
# Get the top 10 hottest summer days
shanghai_hottest = shanghai_summers.sort_values(by=['TEMP', 'HUMI'], ascending=False)
shanghai_hottest[0:10]

season    PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                                
2013 8     9       2.0  67.0  23.2  49.8  1005.1  35.7   8.4            0.0   
           8       2.0  54.2  22.2  47.8  1005.3  35.7  15.8            0.0   
           7       2.0  62.3  23.1  51.6  1005.7  35.4  15.4            0.0   
           6       2.0  43.2  22.6  51.0  1004.2  35.2  18.5            0.0   
     7     29      2.0  39.7  22.7  49.9  1003.9  35.1  11.2            0.0   
     8     10      2.0  45.4  23.3  52.6  1006.0  35.0  23.0            0.0   
     7     25      2.0  49.1  23.1  51.7  1003.1  35.0  16.0            0.0   
           30      2.0  36.0  22.6  51.1  1001.7  34.9  26.1            0.0   
           28      2.0  39.7  22.4  49.4  1004.3  34.9  10.2            0.0   
2010 8     13      2.0   NaN  25.3  59.6  1003.6  34.8  57.2            0.0   

                Iprec  
year month day         
2013 8     9      0.0  
           8      0.0  
           7      0.0  
           6      0.0  
     7     29     0.0  
     8     10     0.0  
     7     25     0.0  
           30     0.0  
           28     0.0  
2010 8     13     0.0

In [33]:
# Filter based on seasons == 4.0 (winter)
shanghai_winter = shanghai_df_by_day['season'] == 4.0
shanghai_winter[325:365]
shanghai_winters = shanghai_df_by_day.loc[shanghai_winter]
shanghai_winters[131:200]

season     PM  DEWP  HUMI    PRES  TEMP    Iws  precipitation  \
year month day                                                                  
2011 2     11      4.0    NaN  -3.4  65.4  1027.0   2.7   33.5            0.0   
           12      4.0    NaN  -5.8  58.8  1029.6   1.6   51.9            0.0   
           13      4.0    NaN  -4.0  68.0  1026.4   1.4   41.8            0.0   
           14      4.0    NaN  -4.9  68.0  1029.6   0.7   18.0            0.0   
           15      4.0    NaN  -6.8  57.2  1031.1   1.1    7.1            0.0   
...                ...    ...   ...   ...     ...   ...    ...            ...   
2012 1     16      4.0   21.2   1.3  72.3  1018.6   6.1  293.6            0.0   
           17      4.0   29.7   0.5  69.1  1021.3   6.0   48.4            0.0   
           18      4.0   35.5   4.0  79.2  1019.5   7.5   69.8            0.4   
           19      4.0  123.0   3.0  77.8  1022.2   6.7  182.7            0.1   
           20      4.0  131.2   0.8  70.3  1023.2   5.9   31.0            0.0   

                Iprec  
year month day         
2011 2     11     0.0  
           12     0.0  
           13     0.0  
           14     0.0  
           15     0.0  
...               ...  
2012 1     16     1.0  
           17     0.0  
           18     2.2  
           19     1.2  
           20     0.0  

[69 rows x 9 columns]

In [34]:
shanghai_coldest = shanghai_winters.sort_values(by='TEMP')
shanghai_coldest[0:10]

season    PM  DEWP  HUMI    PRES  TEMP    Iws  precipitation  \
year month day                                                                 
2011 1     16      4.0   NaN -10.5  55.4  1034.8  -2.5   35.2            0.0   
2010 1     13      4.0   NaN -12.0  45.4  1032.7  -1.5   53.7            0.0   
2013 2     8       4.0  27.9  -3.8  83.7  1033.8  -1.4  147.0            0.0   
2012 12    30      4.0  74.4  -6.3  68.5  1028.6  -0.9  194.0            0.0   
2011 1     17      4.0   NaN  -7.8  60.5  1034.0  -0.8   13.7            0.0   
2013 1     3       4.0  45.9  -4.8  73.9  1038.6  -0.7   49.8            0.0   
2010 2     12      4.0   NaN  -5.7  69.6  1032.0  -0.7   85.4            0.0   
2012 2     8       4.0  45.6  -8.8  55.1  1030.5  -0.6   18.4            0.0   
2010 12    31      4.0   NaN -18.4  25.0  1026.4  -0.5  302.5            0.0   
     1     5       4.0   NaN  -7.0  61.9  1028.7  -0.5   28.5            0.0   

                Iprec  
year month day         
2011 1     16     0.0  
2010 1     13     0.0  
2013 2     8      0.0  
2012 12    30     0.4  
2011 1     17     0.0  
2013 1     3      0.0  
2010 2     12     0.0  
2012 2     8      0.0  
2010 12    31     0.0  
     1     5      0.0

**The hottest summer in Shanghai for this dataset was summer of 2013 because it has the most instances (9) in the top 10 hottest summer days. The coldest winter in Shanghai for this dataset was the winter of 2010 because it has 4 instances in the top 10 coldest winter days, more than any other winter.**

In [35]:
# Filter based on seasons == 2.0 (summer)
shenyang_summer = shenyang_df_by_day['season'] == 2.0
shenyang_summer[195:245]
shenyang_summers = shenyang_df_by_day.loc[shenyang_summer]
shenyang_summers[131:200]

season  PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                              
2011 7     10      2.0 NaN  17.9  67.2  1007.8  25.0  22.4            0.0   
           11      2.0 NaN  18.7  68.0  1009.0  25.4  31.0            0.0   
           12      2.0 NaN  19.3  72.9  1008.9  25.0  25.9            0.0   
           13      2.0 NaN  20.2  72.5  1005.8  26.1   7.4            0.0   
           14      2.0 NaN  20.8  76.8  1006.5  25.6   5.3            0.5   
...                ...  ..   ...   ...     ...   ...   ...            ...   
2012 6     12      2.0 NaN  10.8  60.9   998.6  19.3  11.4            0.4   
           13      2.0 NaN  15.4  79.5  1004.8  19.2  10.1            0.0   
           14      2.0 NaN  16.9  91.5  1007.8  18.3  45.4            0.6   
           15      2.0 NaN  16.2  82.5  1005.6  19.4   6.4            0.3   
           16      2.0 NaN  16.3  84.4  1004.1  19.1  12.1            0.3   

                Iprec  
year month day         
2011 7     10     0.0  
           11     0.0  
           12     0.0  
           13     0.0  
           14     0.7  
...               ...  
2012 6     12     1.0  
           13     0.5  
           14     3.8  
           15     0.5  
           16     0.5  

[69 rows x 9 columns]

In [36]:
# Get the top 10 hottest summer days
shenyang_hottest = shenyang_summers.sort_values(by=['TEMP', 'HUMI'], ascending=False)
shenyang_hottest[0:10]

season    PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                                
2014 8     2       2.0  30.3  22.7  73.5  1003.8  28.6  11.4            0.0   
           3       2.0  31.6  23.3  75.6  1001.0  28.3  11.2            0.0   
2013 8     11      2.0   NaN  24.4  81.4  1008.0  28.1  25.9            0.0   
2014 7     20      2.0  76.1  23.1  74.9  1006.7  28.1  16.4            NaN   
     8     1       2.0  20.7  23.3  76.6  1007.0  28.0  19.1            0.0   
2013 8     7       2.0   NaN  25.5  87.0  1001.4  27.9  14.2            0.8   
           15      2.0   NaN  24.5  83.0  1003.2  27.7  35.4            0.0   
           14      2.0   NaN  23.7  79.8  1003.8  27.6  37.5            0.0   
2012 7     25      2.0   NaN  24.3  82.8  1008.7  27.5  15.7            0.0   
2015 7     9       2.0  52.8  16.9  55.8  1013.1  27.5   7.8            0.0   

                Iprec  
year month day         
2014 8     2      0.0  
           3      0.0  
2013 8     11     0.0  
2014 7     20     NaN  
     8     1      0.0  
2013 8     7      0.8  
           15     0.0  
           14     0.0  
2012 7     25     0.0  
2015 7     9      0.0

In [37]:
# Filter based on seasons == 4.0 (winter)
shenyang_winter = shenyang_df_by_day['season'] == 4.0
shenyang_winter[325:365]
shenyang_winters = shenyang_df_by_day.loc[shenyang_winter]
shenyang_winters[131:200]

season  PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                              
2011 2     11      4.0 NaN -20.6  55.3  1028.5 -13.2  24.2            NaN   
           12      4.0 NaN -17.4  66.8  1031.1 -12.3   5.0            NaN   
           13      4.0 NaN -18.7  62.2  1030.3 -12.6   8.7            NaN   
           14      4.0 NaN -20.3  53.9  1030.3 -12.5  13.9            NaN   
           15      4.0 NaN -15.2  51.6  1024.7  -6.0  26.0            NaN   
...                ...  ..   ...   ...     ...   ...   ...            ...   
2012 1     16      4.0 NaN -12.4  61.3  1020.7  -5.9  10.9            NaN   
           17      4.0 NaN -11.7  80.0  1023.0  -8.8   2.6            NaN   
           18      4.0 NaN -11.1  83.0  1025.5  -8.7   6.9            NaN   
           19      4.0 NaN -12.7  72.1  1028.9  -8.2  19.0            NaN   
           20      4.0 NaN -19.1  50.0  1034.8 -10.7  25.4            NaN   

                Iprec  
year month day         
2011 2     11     NaN  
           12     NaN  
           13     NaN  
           14     NaN  
           15     NaN  
...               ...  
2012 1     16     NaN  
           17     NaN  
           18     NaN  
           19     NaN  
           20     NaN  

[69 rows x 9 columns]

In [38]:
shenyang_coldest = shenyang_winters.sort_values(by='TEMP')
shenyang_coldest[0:10]

season  PM  DEWP  HUMI    PRES  TEMP   Iws  precipitation  \
year month day                                                              
2011 1     16      4.0 NaN -28.0  62.8  1035.7 -22.7   3.9            NaN   
2013 1     2       4.0 NaN -27.9  57.5  1033.0 -21.6   9.7            0.0   
2010 1     3       4.0 NaN -26.0  67.2  1026.8 -21.4   8.1            NaN   
     12    14      4.0 NaN -26.4  64.9  1030.0 -21.4   5.6            NaN   
     1     12      4.0 NaN -26.1  63.8  1024.5 -20.9   4.0            NaN   
2011 1     15      4.0 NaN -27.3  56.2  1036.8 -20.8  10.8            NaN   
2012 12    23      4.0 NaN -27.6  54.6  1033.2 -20.7   4.2            0.0   
2010 1     5       4.0 NaN -24.7  69.4  1029.4 -20.4   2.6            NaN   
           2       4.0 NaN -25.2  64.3  1026.2 -20.1   3.0            NaN   
           6       4.0 NaN -22.9  76.2  1031.2 -19.8   1.8            NaN   

                Iprec  
year month day         
2011 1     16     NaN  
2013 1     2      0.0  
2010 1     3      NaN  
     12    14     NaN  
     1     12     NaN  
2011 1     15     NaN  
2012 12    23     0.0  
2010 1     5      NaN  
           2      NaN  
           6      NaN

**The hottest summer in Chengdu for this dataset was summer of 2014 because although it is tied with 2013 for the most instances (4) in the top 10 hottest summer days, all 4 of its instances are in the top 5. The coldest winter in Chendgu for this dataset was the winter of 2010 because it has 6 instances in the top 10 coldest winter days, more than any other winter.**

**Overall, the hottest summer was the summer of 2010, as it is the hottest summer for 3 of the 5 cities.
The coldest overall winter was winter for 2010 because it is the coldest winter of 3 of the 5 cities.**

**By city, Shanghai summer of 2013 and Guangzhou summer of 2010 were the hottest. Shanghai recorded a 34.9-35.7 temperature range with a 47.8-52.6 humidity range, while Guangzhou recorded a 31.2-32.2 temperature range with a 59.4-66.5 humidity range; the aforementioned summers for both cities had 9 of the 10 hottest summer days, with 8 of Guangzhou's days having humidity levels in the 60s. The coldest winter by city was Shengang winter of 2010 with a temperatures ranging from -21.4 to -19.8.**

- Which feature appears to contain the largest amount of missing data overall?

In [55]:
# Print the total number of null instances for each Beijing feature
for column in beijing_df:
    i = beijing_df[column]
    null_sum = (i.isnull() == True).sum()
    print("Number of null " + column + " instances: ")
    print(null_sum)

Number of null year instances: 
0
Number of null month instances: 
0
Number of null day instances: 
0
Number of null hour instances: 
0
Number of null season instances: 
0
Number of null PM instances: 
2197
Number of null DEWP instances: 
5
Number of null HUMI instances: 
339
Number of null PRES instances: 
339
Number of null TEMP instances: 
5
Number of null Iws instances: 
5
Number of null precipitation instances: 
484
Number of null Iprec instances: 
484


In [56]:
# Print the total number of null instances for each Chengdu feature
for column in beijing_df:
    i = chengdu_df[column]
    null_sum = (i.isnull() == True).sum()
    print("Number of null " + column + " instances: ")
    print(null_sum)

Number of null year instances: 
0
Number of null month instances: 
0
Number of null day instances: 
0
Number of null hour instances: 
0
Number of null season instances: 
0
Number of null PM instances: 
23684
Number of null DEWP instances: 
529
Number of null HUMI instances: 
535
Number of null PRES instances: 
521
Number of null TEMP instances: 
527
Number of null Iws instances: 
533
Number of null precipitation instances: 
2955
Number of null Iprec instances: 
2955


In [57]:
# Print the total number of null instances for each Guangzhou feature
for column in beijing_df:
    i = guangzhou_df[column]
    null_sum = (i.isnull() == True).sum()
    print("Number of null " + column + " instances: ")
    print(null_sum)

Number of null year instances: 
0
Number of null month instances: 
0
Number of null day instances: 
0
Number of null hour instances: 
0
Number of null season instances: 
1
Number of null PM instances: 
20232
Number of null DEWP instances: 
1
Number of null HUMI instances: 
1
Number of null PRES instances: 
1
Number of null TEMP instances: 
1
Number of null Iws instances: 
1
Number of null precipitation instances: 
1
Number of null Iprec instances: 
1


In [58]:
# Print the total number of null instances for each Shanghai feature
for column in beijing_df:
    i = shanghai_df[column]
    null_sum = (i.isnull() == True).sum()
    print("Number of null " + column + " instances: ")
    print(null_sum)

Number of null year instances: 
0
Number of null month instances: 
0
Number of null day instances: 
0
Number of null hour instances: 
0
Number of null season instances: 
0
Number of null PM instances: 
18545
Number of null DEWP instances: 
13
Number of null HUMI instances: 
13
Number of null PRES instances: 
28
Number of null TEMP instances: 
13
Number of null Iws instances: 
12
Number of null precipitation instances: 
4009
Number of null Iprec instances: 
4009


In [59]:
# Print the total number of null instances for each Shenyang feature
for column in beijing_df:
    i = shenyang_df[column]
    null_sum = (i.isnull() == True).sum()
    print("Number of null " + column + " instances: ")
    print(null_sum)

Number of null year instances: 
0
Number of null month instances: 
0
Number of null day instances: 
0
Number of null hour instances: 
0
Number of null season instances: 
0
Number of null PM instances: 
30904
Number of null DEWP instances: 
692
Number of null HUMI instances: 
680
Number of null PRES instances: 
692
Number of null TEMP instances: 
692
Number of null Iws instances: 
692
Number of null precipitation instances: 
12763
Number of null Iprec instances: 
12763


**For each individual city as well as all cities overall, the PM feature appears to contain the largest amount of missing data. Even a city such as Guangzhou, which has very few null values (8 total not including the PM feature), has 20,232 null PM values.**